File: vane-core/src/main/java/org/oddlama/vane/core/config/ConfigStringListMapField.java
package org.oddlama.vane.core.config;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import org.bukkit.configuration.file.YamlConfiguration;
import org.oddlama.vane.annotation.config.ConfigStringListMap;
import org.oddlama.vane.annotation.config.ConfigStringListMapEntry;
import org.oddlama.vane.core.YamlLoadException;

public class ConfigStringListMapField extends ConfigField<Map<String, List<String>>> {

    public ConfigStringListMap annotation;

    public ConfigStringListMapField(
        Object owner,
        Field field,
        Function<String, String> mapName,
        ConfigStringListMap annotation
    ) {
        super(owner, field, mapName, "map of string to string list", annotation.desc());
        this.annotation = annotation;
    }

    // Convert a key like "default" or "terralith_rare" into PascalCase: "Default" / "TerralithRare"
    private static String toPascalCase(final String key) {
        if (key == null || key.isEmpty()) return key;
        final var parts = key.split("[^A-Za-z0-9]+");
        final var sb = new StringBuilder();
        for (final var p : parts) {
            if (p.isEmpty()) continue;
            sb.append(Character.toUpperCase(p.charAt(0)));
            if (p.length() > 1) sb.append(p.substring(1));
        }
        return sb.toString();
    }

    // Normalize keys from YAML back to canonical internal form (lowercase, non-alphanum -> underscore)
    private static String normalizeKey(final String key) {
        if (key == null) return null;
        return key.toLowerCase().replaceAll("[^a-z0-9]+", "_");
    }

    private void appendStringListMapDefinition(
        StringBuilder builder,
        String indent,
        String prefix,
        Map<String, List<String>> def
    ) {
        def.forEach((k, list) -> {
            builder.append(indent);
            builder.append(prefix);
            builder.append("  ");
            // Use PascalCase for keys in generated YAML
            builder.append(escapeYaml(toPascalCase(k)));
            builder.append(":\n");

            list.forEach(s -> {
                builder.append(indent);
                builder.append(prefix);
                builder.append("    - ");
                builder.append(escapeYaml(s));
                builder.append("\n");
            });
        });
    }

    @Override
    public Map<String, List<String>> def() {
        final var override = overriddenDef();
        if (override != null) {
            return override;
        } else {
            return Arrays.stream(annotation.def()).collect(
                Collectors.toMap(ConfigStringListMapEntry::key, e -> Arrays.asList(e.list()))
            );
        }
    }

    @Override
    public boolean metrics() {
        final var override = overriddenMetrics();
        if (override != null) {
            return override;
        } else {
            return annotation.metrics();
        }
    }

    @Override
    public void generateYaml(StringBuilder builder, String indent, YamlConfiguration existingCompatibleConfig) {
        appendDescription(builder, indent);

        // Default
        builder.append(indent);
        builder.append("# Default:\n");
        appendStringListMapDefinition(builder, indent, "# ", def());

        // Definition
        builder.append(indent);
        builder.append(basename());
        builder.append(":\n");
        final var def = existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath())
            ? loadFromYaml(existingCompatibleConfig)
            : def();
        appendStringListMapDefinition(builder, indent, "", def);
    }

    @Override
    public void checkLoadable(YamlConfiguration yaml) throws YamlLoadException {
        checkYamlPath(yaml);

        if (!yaml.isConfigurationSection(yamlPath())) {
            throw new YamlLoadException("Invalid type for yaml path '" + yamlPath() + "', expected group");
        }

        for (var listKey : yaml.getConfigurationSection(yamlPath()).getKeys(false)) {
            final var listPath = yamlPath() + "." + listKey;
            if (!yaml.isList(listPath)) {
                throw new YamlLoadException("Invalid type for yaml path '" + listPath + "', expected list");
            }

            for (var obj : yaml.getList(listPath)) {
                if (!(obj instanceof String)) {
                    throw new YamlLoadException("Invalid type for yaml path '" + listPath + "', expected string");
                }
            }
        }
    }

    public Map<String, List<String>> loadFromYaml(YamlConfiguration yaml) {
        final var map = new HashMap<String, List<String>>();
        for (final var listKey : yaml.getConfigurationSection(yamlPath()).getKeys(false)) {
            final var listPath = yamlPath() + "." + listKey;
            final var list = new ArrayList<String>();
            // Normalize keys so in-memory representation stays lowercase/underscore
            map.put(normalizeKey(listKey), list);
            for (final var obj : yaml.getList(listPath)) {
                list.add((String) obj);
            }
        }
        return map;
    }

    public void load(YamlConfiguration yaml) {
        try {
            field.set(owner, loadFromYaml(yaml));
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Invalid field access on '" + field.getName() + "'. This is a bug.");
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/ConfigStringListMapField.kt
package org.oddlama.vane.core.config

import org.bukkit.configuration.file.YamlConfiguration
import org.oddlama.vane.annotation.config.ConfigStringListMap
import org.oddlama.vane.core.YamlLoadException
import java.lang.reflect.Field
import java.util.Locale

class ConfigStringListMapField(
    owner: Any?,
    field: Field,
    mapName: (String?) -> String?,
    var annotation: ConfigStringListMap
) : ConfigField<MutableMap<String?, MutableList<String?>?>?>(
    owner, field, mapName, "map of string to string list", annotation.desc
) {
    private fun appendStringListMapDefinition(builder: StringBuilder, indent: String?, prefix: String?, def: MutableMap<String?, MutableList<String?>?>) {
        def.forEach { (k, list) ->
            builder.append("$indent$prefix  ${escapeYaml(toPascalCase(k))}:\n")
            list!!.forEach { s -> builder.append("$indent$prefix    - ${escapeYaml(s)}\n") }
        }
    }

    override fun def(): MutableMap<String?, MutableList<String?>?> =
        overriddenDef() ?: annotation.def.associate { e -> e.key to e.list.toMutableList<String?>() }.toMutableMap()

    override fun metrics(): Boolean = overriddenMetrics() ?: annotation.metrics

    override fun generateYaml(builder: StringBuilder, indent: String, existingCompatibleConfig: YamlConfiguration?) {
        appendDescription(builder, indent)
        builder.append("$indent# Default:\n")
        appendStringListMapDefinition(builder, indent, "# ", def())
        builder.append("$indent${basename()}:\n")
        val def = if (existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath()))
            loadFromYaml(existingCompatibleConfig) else def()
        appendStringListMapDefinition(builder, indent, "", def)
    }

    @Throws(YamlLoadException::class)
    override fun checkLoadable(yaml: YamlConfiguration) {
        checkYamlPath(yaml)
        if (!yaml.isConfigurationSection(yamlPath()))
            throw YamlLoadException("Invalid type for yaml path '${yamlPath()}', expected group")
        for (listKey in yaml.getConfigurationSection(yamlPath())!!.getKeys(false)) {
            val listPath = "${yamlPath()}.$listKey"
            if (!yaml.isList(listPath))
                throw YamlLoadException("Invalid type for yaml path '$listPath', expected list")
            for (obj in yaml.getList(listPath)!!) {
                if (obj !is String)
                    throw YamlLoadException("Invalid type for yaml path '$listPath', expected string")
            }
        }
    }

    fun loadFromYaml(yaml: YamlConfiguration): MutableMap<String?, MutableList<String?>?> =
        yaml.getConfigurationSection(yamlPath())!!.getKeys(false).associateTo(mutableMapOf()) { listKey ->
            val listPath = "${yamlPath()}.$listKey"
            normalizeKey(listKey) to yaml.getList(listPath)!!.map { it as String? }.toMutableList()
        }

    override fun load(yaml: YamlConfiguration) {
        try {
            field.set(owner, loadFromYaml(yaml))
        } catch (_: IllegalAccessException) {
            throw RuntimeException("Invalid field access on '${field.name}'. This is a bug.")
        }
    }

    companion object {
        private fun toPascalCase(key: String?): String? {
            if (key.isNullOrEmpty()) return key
            return key.split("[^A-Za-z0-9]+".toRegex())
                .filter { it.isNotEmpty() }
                .joinToString("") { it[0].uppercaseChar() + it.substring(1) }
        }

        private fun normalizeKey(key: String?): String? =
            key?.lowercase(Locale.getDefault())?.replace("[^a-z0-9]+".toRegex(), "_")
    }
}
